Lib/test/test_fractions.py

```diff
@@ -346,6 +346,34 @@ def testConversions(self):
 
         self.assertTypedEquals(0.1+0j, complex(F(1,10)))
 
+    def testBoolGuarateesBoolReturn(self):
+        # Ensure that __bool__ is used on numerator which guarantees a bool
+        # return.  See also bpo-39274.
+        class custom_value(int):
+            def __init__(self, value):
+                self.value = value
+
+            def __bool__(self):
+                self.bool_used = True
+                return bool(self.value)
+
+            @property
+            def numerator(self):
+                # required to preserve `self` during instantiation
+                return self
+
+        numerator = custom_value(1)
+        r = F(numerator)
```

Comment:
I would prefer to explicit the denominator: F(numerator, 1). Same below.

Reply:
That does not change much, if I add that, the only difference is that I also have to define `__mul__`. The only solution is going back to `r._numerator = ...` (it is used elsewhere in the file)